js/src/components/shipping-rate-section/minimum-order-card/changeMinimumOrderGroup.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * @typedef { import(".~/data/actions").ShippingRate } ShippingRate
+ * @typedef { import("./typedefs").MinimumOrderGroup } MinimumOrderGroup
+ */
+
+/**
+ * Returns new shipping rates value, with updated group.
+ *
+ * Will set `shipingRate.options.free_shipping_threshold` to `newGroup.threshold`
+ * for the shipping rate countries that are listed in `newGroup.countries` (if provided),
+ * and set `threshold` to `undefined` for the countries which were in `oldGroup.countries` (if provided),
+ * but not in the new.
+ *
+ * @param {Array<ShippingRate>} value Shipping rates.
+ * @param {MinimumOrderGroup} [oldGroup] Old minimum order group.
+ * @param {MinimumOrderGroup} [newGroup] New minimum order group.
+ * @return {Array<ShippingRate>} New, updated value.
+ */
+export const changeMinimumOrderGroup = ( value, oldGroup, newGroup ) => {
```

Comment:
The hardest part of programming - naming ;)

I'm also not happy with `changeMinimumOrderGroup` for the exact reason you mentioned. However, I also don't like `getShippingRateNewValue`:
- It sounds like a getter, so why would I need to pass any args?
- Does it always get up-to-date value?
- `getShippingRateNewValue( value, oldGroup, newGroup ) `:
   - To get a value I need to give a value?
   - "Group" what does a group have to do for shipping rate value? Is that a shipping rates group?
   
 The fact it returns a new value instead of changing the existing one, does not look extremely important to emphasize. Given we write in a framework that tries to embrace functional programming and immutability. (If we already pay a price for this, let's take some benefits ;) )
   
   Personally, I like when the function name is optimized for a consumer's perspective, not implementer's perspective. So instead of stating what the function technically does (like the source code do), it gives hints for the consumer, like:
- What does it do conceptually?
- When would I need/ use it? 

So I would describe this function as: "changeShippingRatesArrayAcordingToChangedMinimumOrderGroup", "changeMinimumOrderOfAGroupWithinShippingRates", "applyMiniumOrderGroupChangeToShippingRates" - but those names are a bit lengthy :joy: 

----


Maybe we can put it that way:

`changeMinimumOrderGroup` is a small utility function, to be used here locally, not plugin wide. This is util used only in `MinimumOrderCard` defined and used only within `ShippingRateSection`. Therefore, maybe we can assume that from the perspective, of the only consumer of this function - `MinimunOrderCard`:
- the "value" is always "shipping rates array" (as specified in props) 
- "group"/"countries group/threshold group" is always "Mimimum Order Group"? (as defined in [code](https://github.com/woocommerce/google-listings-and-ads/pull/1499/files/3109ee4f7f8d5b0ba6d67f6ff8f1df54df1ab7ca#diff-19741a9dbe0895f99f4b04a6a88e21d3a8445917bea2f62ccf3f2610acff92d8R31))
Then we can call it:
`changeGroupWithinValue( value, oldGroup, newGroup )`?

@ecgan , @puntope WDYT?

Reply:
> However, I also don't like `getShippingRateNewValue`: [...]

Interesting. For me, I usually just use `get*` as a short and simple way to name a function that returns a value, and it is fine to have args. Alternative naming can be `calculate*`, `calc*`, `generate*`. I also tend to think of these `get*` functions as pure functions that don't mutate the args and can be easily tested with unit tests.

> Personally, I like when the function name is optimized for a _consumer's perspective_, not implementer's perspective. [...]

Yes, agree. I just want to be more specific and say it is for the function code consumer perspective (not the UI end-user consumer perspective; for some reasons I felt the naming was more towards UI end-user perspective, probably because of the word "change" that relates to the users changing the minimum order group in the UI).

---

> Maybe we can put it that way: [...]

Yes, totally agree with those points. 😄 👍 

> * Then we can call it:
>   `changeGroupWithinValue( value, oldGroup, newGroup )`?

I still don't quite like the "changeGroup" naming. From the code consumer perspective, I feel it is like it is going to mutate `oldGroup` or `newGroup` and not going to return a value. 😅 

Anyway, I think if we can't reach an agreement here, we can move on with we have, either `changeMinimumOrderGroup` or `changeGroupWithinValue` is fine and it doesn't make a big difference.

Reply:
>  usually just use `get*` as a short and simple way to name a function that returns a value, and it is fine to have args. Alternative naming can be `calculate*`, `calc*`, `generate*`

Personally, I like when most functions return some value ;) and to me `get*` is more connected to "obtaining" something - like in `localStorage.getItem`, than just returning something. Returning something is a natural characteristic of a function. And here we are not obtaining/accessing/fetching anything.
We are doing state management translation. Having the original state in "local" API and two states in "remote"/"child" API, we rather translate that change and apply it to a given value. So, we either `applyGroupChange`, or  `calculate*` :smiley: the new state out of other states.

> it is for the function code consumer perspective (not the UI end-user consumer perspective;

Good point, to focus more on code customer perspective rather than end-user using UI. (However, I think the latter also has some value. For example new engineers, not familiar with the codebase, who knows what the end product looks like and what is for, but don't know the code yet. They may not know how the state is calculated and what are data types, but they know that the component is meant to change the minimum order for a given group of countries).

Also, as mentioned above, from the code perspective there is also a change. The initial state of the group changed to a new one. Then, this function translates the state change of `MinimumOrderGroup` to the state change in `Array<ShippingRates>`.
But taking the perspective and _"I feel it is like it is going to mutate […]"_ points, I now lean more towards `apply*` or `calculate*`. As for our use case, we do not need to change or mutate anything, we need to return an object with the new state.

WDYT of `applyGroupChange`? 
`calculateValue` does not give much information from what data/in what case/when, so I'd use `calcuateValueFromGroupChange` - which is lengthy

----

But...
As we control both parties: `MinimumOrderCard` and `MinimumOrderInputControl` instead of trying to find a perfect name for a function that translates one API/data type to another, we could make both components use the same API/data structure. Then the problem of naming that function, implementing it, and testing will be simply gone. :smile:

Reply:
> to me `get*` is more connected to "obtaining" something - like in `localStorage.getItem`, than just returning something.

Interesting. I guess you mean `get*` is to get something that already exists somewhere (e.g. localstorage, API, database etc, the things are there, instead of generated / calculated / created in the function itself). Thanks, good to know, and I think / hope I can get use to that 🙂 👍 

> `calculateValue` does not give much information from what data/in what case/when, so I'd use `calculateValueFromGroupChange` - which is lengthy

I'd say both of them are better than `changeMinimumOrderGroup` 😄 👍 Personally I would go for `calculateValue`, because by looking at the function parameters with `oldGroup` and `newGroup`, I would already know it is associated with "group change", so having "*FromGroupChange" doesn't help much for me. But honestly I don't mind `calculateValueFromGroupChange` either, it is not the longest function name in the code base. 😆 We can go with `calculateValueFromGroupChange`.

---

> As we control both parties: `MinimumOrderCard` and `MinimumOrderInputControl` instead of trying to find a perfect name for a function that translates one API/data type to another, we could make both components use the same API/data structure. Then the problem of naming that function, implementing it, and testing will be simply gone. 😄

Hmm, I don't quite get this. I mean, no matter where in the code base, we still need to do that data transformation somewhere, to transform / translate shipping rate data to/from minimum order group data. 🤔 

Maybe a separate PR one day, love to see your idea. 😄

Reply:
I forgot about the discussion we have here and merged this PR after I merged PR #1410. I have created PR https://github.com/woocommerce/google-listings-and-ads/pull/1513 to continue the function naming change and discussion. 🙏